docs/en/sql-reference/statements/create/table.md

```diff
@@ -380,7 +380,7 @@ High compression levels are useful for asymmetric scenarios, like compress once,
 
 `DEFLATE_QPL` — [Deflate compression algorithm](https://github.com/intel/qpl) implemented by Intel® Query Processing Library. Some limitations apply:
 
-- DEFLATE_QPL is experimental and can only be used after setting configuration parameter `allow_experimental_codecs=1`.
+- DEFLATE_QPL is disabled by default and can only be used after setting configuration parameter `enable_qpl_deflate=1`.
```

Comment:
There are two types of settings in ClickHouse: user/session/query-level settings and server-level (aka. global) settings. The former are centrally defined in Core/Settings.h, documented [here](https://clickhouse.com/docs/en/operations/settings/settings) and exist at runtime in the non-shared part of the `Context` (Interpreters/Context.h). The latter are not centrally defined, but documented [here](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings), and they exist in the shared part of the `Context`.

The previously used `allow_experimental_codecs = 1` and the new parameter `enable_qpl_deflate` are user settings. The disadvantage with them is that they need to be set anew in every session using `SET enable_qpl_deflate = 1;` (or alternatively using some unbeautiful syntax in config file users.xml) - this is a bit annoying and unintuitive for something that should generally work. So, a better alternative would be to control QPL on&off at server-level. Perhaps that is something which can be done in a separate PR.

Reply:
Noted it. I think I should implement a new setting like below:

<compression>
    <case>
      <min_part_size>...</min_part_size>
      <min_part_size_ratio>...</min_part_size_ratio>
      <method>deflate_qpl</method>
      <enable_qpl_deflate_codec>1</enable_qpl_deflate_codec>
    </case>
    ...
</compression>

If "enable_qpl_deflate_codec" to be set 0 or not defined here , the server would throw an exception.
And the code change focus on this file: ./src/Storages/CompressionCodecSelector.h
If my understanding above is correct, let me try to implement it on this PR. @rschu1ze

Reply:
Yepp, your understanding is correct. It would be great to have this implemented in this PR - in this case, we would avoid an intermediate code state where QPL can be enabled/disabled via user-defined settings.

The rest of the PR looks good (I did some minor adjustments), thanks.

Reply:
New commits added. But the current situation is that deflate_qpl codec can be controlled from server setting perspective, while user still can use it through customizing codec like below:
```
create table test_array (`ab_version` Int32 CODEC(DEFLATE_QPL)) Engine = Log();
insert into test_array select * from numbers(1000);
```
Is that a problem?

Reply:
It is indeed still possible to specify the codec manually. The reason is that the new check in Storages/CompressionCodecSelector.h only affects a heuristics which automatically applies compression to data parts with specific properties (i.e. a minimum size, a minimum part-size ratio etc.).

So yes, enablement/disablement would need to be more comprehensive since explicit specification of the codec is common practice in ClickHouse. (sorry if I told you wrong ... this was not clear to me).

What about this:
- you add a check similar to what we had before (see [here](https://github.com/ClickHouse/ClickHouse/pull/50305/commits/1afb7fbce5468c12055511741883d224b51897b2)) in `validateCodecAndGetPreprocessedAST`. But instead of passing in a parameter `enable_qpl_default` as argument (which eventually came from `context->getSettingsRef().enable_qpl_deflate_codec`), we parse the server configuration file (like done in Storages/CompressionCodecSelector.h). This approach should capture all situations where a codec is created.
- I think that with this, we no longer need to change `CompressionCodecSelector`.

I also noticed that `<enable_qpl_deflate_codec>` is nested in <case>. That makes little sense, it should be defined directly below `<compression>`.

Reply:
New commits added. Two more questions:
1. We cannot define enable_qpl_deflate_codec inside the section of<compression>, because right now it limit the keyword start with "case" since the logic need loop the conditions start with "case", otherwise exception would happen. So i create a separate server setting named "enable_deflate_qpl_codec" which stand the same level with "compression".
![image](https://github.com/ClickHouse/ClickHouse/assets/19898583/e1f313ba-c4cd-4f5e-b381-d8446864fcb7)

2.  The config context cannot be shared in below function, which actually called from client side, but it validate codec as well. I have to pass the value of enable_deflate_qpl_codec with true, is that a problem?
Otherwise, we need to make more changes on caller function : Connection::sendQuery
./src/Client/Connection.cpp  --> Connection::sendQuery
`CompressionCodecFactory::instance().validateCodec(method, level, !settings->allow_suspicious_codecs, settings->allow_experimental_codecs, true);`

Reply:
1. isn't a question 😄 but the approach you chose is the right one!
2. Should be fine to pass `true` as dummy value (I'll validate locally that everything behaves as expected)

Reply:
One more question, since we need enable deflate_qpl on server side , how to run function test in CI system?
On local test, i can specify config.xml with enable_deflate_qpl_codec=true, but in CI system, is there any specific server config need to change?